### DISCUSSION TITLE
ZuluSCSI RP2040 mini: ISO image is read-only – failed to load

### DISCUSSION THREAD

**Discussion by ufkel** [ZuluSCSI/ZuluSCSI-firmware#272] [Q&A] [answered]
Hello,

the new ZuluSCSI has just arrived, and up to now it works great with disk images. But with CDROM images, I run into a problem:
ZuluSCSI reports the ISO image is read-only, and ignores it afterwards. See the log details at 130ms.

This is unexpected. Did I do something wrong? The ISO image does work with MacSD, so I think the file is OK.

Many thanks for any idea.

Udo

`

> [10ms] Platform: ZuluSCSI RP2040
> [10ms] FW Version: 23.07.01-release Jun 30 2023 16:12:07
> [11ms] DIP switch settings: debug log 0, termination 1
> [11ms] SCSI termination is enabled
> [12ms] Flash chip size: 2048 kB
> [14ms] SCSI target/disk mode selected by DIP switch, acting as a SCSI disk
> [27ms] SD card detected, FAT64 volume size: 15240 MB
> [28ms] SD MID: 0x74, OID: 0x4A 0x60
> [28ms] SD Name: USDU1
> [29ms] SD Date: 10/2017
> [29ms] SD Serial: 0x42B31AB2
> [30ms] Reading configuration from zuluscsi.ini
> [31ms] Active configuration (using system preset "Mac"):
> [33ms] -- SelectionDelay = 255
> [33ms] -- EnableUnitAttention = No
> [34ms] -- EnableSCSI2 = Yes
> [35ms] -- EnableSelLatch = No
> [36ms] -- MapLunsToIDs = No
> [36ms] -- EnableParity = Yes
> [116ms] -- Opening 'Apple_Legacy_Recovery.iso' for id:5, specified in zuluscsi.ini
> [130ms] ---- Image file is read-only, writes disabled
> [130ms] ---- Failed to load image 'Apple_Legacy_Recovery.iso', ignoring
> [158ms] Finding HDD images in directory /:
> [162ms] -- Opening /HD0.hda for id:0 lun:0
> [166ms] DBG ---- Image file is contiguous, SD card sectors 22848 to 2119999
> [167ms] ---- Read prefetch enabled: 8192 bytes
> [168ms] -- Opening /HD1.hda for id:1 lun:0
> [171ms] DBG ---- Image file is contiguous, SD card sectors 2120000 to 8411455
> [172ms] ---- Read prefetch enabled: 8192 bytes
> [174ms] -- Opening /HD3.hda for id:3 lun:0
> [178ms] DBG ---- Image file is contiguous, SD card sectors 8412608 to 12606910
> [179ms] ---- Read prefetch enabled: 8192 bytes
> [180ms] Finding HDD images in additional directory Dir1 = "/ISO_Images":
> [187ms] -- Platform supports ROM drive up to 1692 kB
> [187ms] ---- ROM drive image not detected
> [188ms] SCSI ID:0 BlockSize:512 Type:0 Quirks:1 ImageSize:1048576kB
> [188ms] SCSI ID:1 BlockSize:512 Type:0 Quirks:1 ImageSize:3145728kB
> [189ms] SCSI ID:3 BlockSize:512 Type:0 Quirks:1 ImageSize:2097151kB
> [290ms] Initialization complete!`

**Comment by aperezbios** [comment]
@PetteriAimonen any thoughts here? @ufkel I'm wondering if the read-only bit is set on the image file itself. Is it? This could be a bug.

**Reply by ufkel** [reply]
Take the very same file, rename it to CD4.iso, it works OK. The problem only appears in conjunction with the CD changer feature IMG0-IMG9.
BTW: The changer function should work even with a single image - right? So there is only IMG0 specified, no IMG1, IMG2, ...

**Reply by PetteriAimonen** [reply]
It's failing to open the image for some reason. Wrong filename would be the obvious reason, but I guess it finds something because of the read-only report.

And yeah, it should work even with just IMG0.

**Reply by phe78** [reply]
i had the same problem with an E-MU E6400 sampler. The problem came from some characters in the file name that were not recognized. I identified this as the Zulu log file displayed the file name with "?' at the place of what looked like a usual dash (it was not). Changed those characters with regular dashes did the trick. Hope that helps.

**Comment by ufkel** [comment]
To me it seems to be a more general problem.

Ignore that special ISO file for a moment. Let's take completely different ones. 
Here is my ```zuluscsi.ini``` file:
```
[SCSI]
# Settings that apply to all SCSI IDs

# Select a system preset to apply default settings
# Known systems: "Mac", "MacPlus"
System="Mac"

Debug = 1   # 1: Same effect as DIPSW2, enables verbose log messages
#SelectionDelay = 255   # Millisecond delay after selection, 255 = automatic, 0 = no delay
#Dir1 = "/"   # Optionally look for image files in subdirectory
#Dir1 = "/ISO_Images"  # Multiple directories can be specified Dir1...Dir9
#DisableStatusLED 1 # 0: Use status LED, 1: Disable status LED

# NOTE: PhyMode is only relevant for ZuluSCSI V1.1 at this time.
#PhyMode = 0   # 0: Best available  1: PIO  2: DMA_TIMER  3: GREENPAK_PIO   4: GREENPAK_DMA
# The PhyMode parameter has no effect on ZuluSCSI RP2040-based platforms, as there is only one PHY mode.

# Settings that can be needed for compatibility with some hosts
#Quirks = 0   # 0: Standard, 1: Apple, 2: OMTI, 4: Xebec, 8: VMS
#EnableUnitAttention = 0 # Post UNIT_ATTENTION status on power-on or SD card hotplug
#EnableSCSI2 = 1 # Enable faster speeds of SCSI2
#EnableSelLatch = 0 # For Philips P2000C and other devices that release SEL signal before BSY
#EnableParity = 1 # Enable parity checks on platforms that support it (RP2040)
#MapLunsToIDs = 0 # For Philips P2000C simulate multiple LUNs
#MaxSyncSpeed = 10 # Set to 5 or 10 to enable synchronous SCSI mode, 0 to disable
#InitPreDelay = 0  # How many milliseconds to delay before the SCSI interface is initialized
#InitPostDelay = 0 # How many milliseconds to delay after the SCSI interface is initialized

# ROM settings
#DisableROMDrive = 1 # Disable the ROM drive if it has been loaded to flash
#ROMDriveSCSIID = 7 # Override ROM drive's SCSI ID

# Settings that can be specified either per-device or for all devices.
#Vendor = "QUANTUM"
#Product = "FIREBALL1"
#Version = "1.0"
#Serial = "0123456789ABCDEF"
#Type = 0     # 0: Fixed, 1: Removable, 2: Optical, 3: Floppy, 4: Mag-optical, 5: Tape
#TypeModifier = 0  # Affects only INQUIRY response
#SectorsPerTrack = 63
#HeadsPerCylinder = 255
#RightAlignStrings = 0 # Right-align SCSI vendor / product strings, defaults on if Quirks = 1
#PrefetchBytes = 8192 # Maximum number of bytes to prefetch after a read request, 0 to disable
#ReinsertCDOnInquiry = 1 # Reinsert any ejected CD-ROM image on Inquiry command
#ReinsertAfterEject = 1 # Reinsert next CD image after eject, if multiple images configured.
#EjectButton = 0 # Enable eject by button 1 or 2, or set 0 to disable

# Settings can be overridden for individual devices.
#[SCSI2]
#Product = "Disk with ID2"

[SCSI5]
Product = "CD-ROM Drive"
Type = 2
# If IMG0..IMG9 are specified, they are cycled after each CD eject command.
IMG0 = "devcdjan92.iso"
IMG1 = "E-T-O.iso"
#IMG1 = SecondCD.bin

# Raw sector range from SD card can be passed through
# Format is RAW:first_sector:last_sector where sector numbers can be decimal or hex.
# If end sector is beyond end of SD card, it will be adjusted automatically.
# [SCSI4]
# IMG0 = RAW:0x00000000:0xFFFFFFFF # Whole SD card
```

Taking aside all the comments, it boils down to 
```
[SCSI]
System="Mac"
Debug = 1 
[SCSI5]
Product = "CD-ROM Drive"
Type = 2
IMG0 = "devcdjan92.iso"
IMG1 = "E-T-O.iso"
```
Here is the directory listing of the SD-card:
```
ukeller /Volumes/ZuluSCSI:ls -l
total 17879616
-rwxrwxrwx@ 1 ukeller  staff   679405568 Aug  7  2019 Apple Legacy Recovery Oct 1999.iso
-rwxrwxrwx@ 1 ukeller  staff   679405568 Jan  1  1980 Cx4.iso
-rwxrwxrwx@ 1 ukeller  staff   681574400 Aug  7  2019 E-T-O.iso
-rwxrwxrwx@ 1 ukeller  staff  1073741824 Jul 24 17:02 HD0.hda
-rwxrwxrwx@ 1 ukeller  staff  3221225472 Jul 24 17:03 HD1.hda
-rwxrwxrwx@ 1 ukeller  staff  2147483136 Jul 24 17:06 Hx3.hda
drwxrwxrwx  1 ukeller  staff       32768 Jul 30 16:59 ISO_Images
-rwxrwxrwx@ 1 ukeller  staff   668436480 May  5  2014 devcdjan92.iso
-rwxrwxrwx@ 1 ukeller  staff     2968150 Dec 31  2022 zululog.txt
-rwxrwxrwx@ 1 ukeller  staff        3007 Aug  2 13:47 zuluscsi.ini
ukeller /Volumes/ZuluSCSI:
```

The SE/30 boots happily from HD0 (as expected), shows HD1 on the desktop (as expected). But it does not show the CDROM volume (not expected).

All SCSI devices are available...
![IMG_6478](https://github.com/ZuluSCSI/ZuluSCSI-firmware/assets/68097348/dd55d3aa-2e98-4d1a-9158-34270d8e5239)
but the CDROM-volume is not
![IMG_6480](https://github.com/ZuluSCSI/ZuluSCSI-firmware/assets/68097348/18e64722-8518-4f16-93bc-c3181fa7fe99)

Here is the beginning of the debug output:
```
[10ms] Platform: ZuluSCSI RP2040
[10ms] FW Version: 23.07.01-release Jun 30 2023 16:12:07
[11ms] DIP switch settings: debug log 0, termination 1
[11ms] SCSI termination is enabled
[12ms] Flash chip size: 2048 kB
[14ms] SCSI target/disk mode selected by DIP switch, acting as a SCSI disk
[27ms] SD card detected, FAT64 volume size: 15240 MB
[28ms] SD MID: 0x74, OID: 0x4A 0x60
[28ms] SD Name: USDU1
[29ms] SD Date: 10/2017
[29ms] SD Serial: 0x42B31AB2
[31ms] Reading configuration from zuluscsi.ini
[31ms] Active configuration (using system preset "Mac"):
[33ms] -- SelectionDelay = 255
[33ms] -- EnableUnitAttention = No
[34ms] -- EnableSCSI2 = Yes
[35ms] -- EnableSelLatch = No
[36ms] -- MapLunsToIDs = No
[36ms] -- EnableParity = Yes
[116ms] -- Opening 'devcdjan92.iso' for id:5, specified in zuluscsi.ini
[129ms] DBG ---- Image file is contiguous, SD card sectors 17960640 to 19266179
[130ms] ---- Read prefetch enabled: 8192 bytes
[158ms] Finding HDD images in directory /:
[162ms] -- Opening /HD0.hda for id:0 lun:0
[165ms] DBG ---- Image file is contiguous, SD card sectors 22848 to 2119999
[166ms] ---- Read prefetch enabled: 8192 bytes
[167ms] -- Opening /HD1.hda for id:1 lun:0
[170ms] DBG ---- Image file is contiguous, SD card sectors 2120000 to 8411455
[171ms] ---- Read prefetch enabled: 8192 bytes
[179ms] -- Platform supports ROM drive up to 1692 kB
[179ms] ---- ROM drive image not detected
[180ms] SCSI ID:0 BlockSize:512 Type:0 Quirks:1 ImageSize:1048576kB
[180ms] SCSI ID:1 BlockSize:512 Type:0 Quirks:1 ImageSize:3145728kB
[181ms] SCSI ID:5 BlockSize:2048 Type:2 Quirks:1 ImageSize:652770kB
[282ms] Initialization complete!
[8375ms] DBG BUS RESET
[8859ms] DBG -- BUS_BUSY
[8859ms] DBG -- BUS_FREE
[9112ms] DBG -- BUS_BUSY
[9112ms] DBG ---- SELECTION: 5
[9114ms] DBG ---- COMMAND: Read6
[9114ms] DBG ------ OUT: 0x08 0x00 0x00 0x00 0x01 0x00 
[9115ms] DBG ------ Read 1x2048 starting at 0
[9115ms] DBG ---- DATA_IN
[9161ms] DBG ---- Total IN: 2048 OUT: 0 CHECKSUM: 44821
[9162ms] DBG ---- STATUS: 0 GOOD
[9180ms] DBG ---- MESSAGE_IN
[9180ms] DBG ------ IN: 0x00 
[9180ms] DBG -- BUS_FREE
[9433ms] DBG -- BUS_BUSY
[9434ms] DBG -- BUS_FREE
[9686ms] DBG -- BUS_BUSY
[9687ms] DBG -- BUS_FREE
[9939ms] DBG -- BUS_BUSY
[9939ms] DBG ---- SELECTION: 1
[9941ms] DBG ---- COMMAND: Read6
[9941ms] DBG ------ OUT: 0x08 0x00 0x00 0x00 0x01 0x00 
[9942ms] DBG ------ Read 1x512 starting at 0
[9943ms] DBG ---- DATA_IN
[9944ms] DBG ---- Total IN: 512 OUT: 0 CHECKSUM: 65527
[9945ms] DBG ---- STATUS: 0 GOOD
[9961ms] DBG ---- MESSAGE_IN
[9962ms] DBG ------ IN: 0x00 
[9962ms] DBG -- BUS_FREE
[9972ms] DBG -- BUS_BUSY
[9973ms] DBG ---- SELECTION: 1
[9974ms] DBG ---- COMMAND: Read6
[9975ms] DBG ------ OUT: 0x08 0x00 0x00 0x65 0x0B 0x00 
[9975ms] DBG ------ Read 11x512 starting at 101
[9976ms] DBG ---- DATA_IN
[9987ms] DBG ---- Total IN: 5632 OUT: 0 CHECKSUM: 50828
[9987ms] DBG ---- STATUS: 0 GOOD
[9997ms] DBG ---- MESSAGE_IN
[9998ms] DBG ------ IN: 0x00 
[9998ms] DBG -- BUS_FREE

[... many more lines ...]
```

When I take those ISO files, and define them as single drives (aka CD4.iso, CD5.iso), everything is OK, the volumes are visible on the desktop. Therefore I assume the ISO files are OK. But as soon as I define them as part of the CDROM changer feature, something is not OK.

**Reply by ufkel** [reply]
I forgot to mention:
ZuluSCSI mini is the only device on the SCSI bus, directly attached to the back of the Macintosh.

**Comment by ufkel** [comment]
Good news!

Once the
```
System="Mac"
```
config line is disabled, the CD changer feature works without a problem.

Now I have to narrow down which of those settings is responsible for the problem:
```
[31ms] Active configuration (using system preset "Mac"):
[33ms] -- SelectionDelay = 255
[33ms] -- EnableUnitAttention = No
[34ms] -- EnableSCSI2 = Yes
[35ms] -- EnableSelLatch = No
[36ms] -- MapLunsToIDs = No
[36ms] -- EnableParity = Yes
```
Stay tuned.

**Comment by ufkel** [comment]
According to the debug log there is no difference in the global settings.

System preset: Mac:
```
[31ms] Active configuration (using system preset "Mac"):
[33ms] -- SelectionDelay = 255
[33ms] -- EnableUnitAttention = No
[34ms] -- EnableSCSI2 = Yes
[35ms] -- EnableSelLatch = No
[36ms] -- MapLunsToIDs = No
[36ms] -- EnableParity = Yes
```
and without a preset system:
```
[91ms] Active configuration:
[92ms] -- SelectionDelay = 255
[93ms] -- EnableUnitAttention = No
[94ms] -- EnableSCSI2 = Yes
[95ms] -- EnableSelLatch = No
[96ms] -- MapLunsToIDs = No
[96ms] -- EnableParity = Yes
```

But there is a difference in the device-specific "quirks" settings.
System preset: Mac:
```
[181ms] SCSI ID:5 BlockSize:2048 Type:2 Quirks:1 ImageSize:652770kB
```
without a preset system:
```
[234ms] SCSI ID:5 BlockSize:2048 Type:2 Quirks:0 ImageSize:652770kB
```

Summary:
On my system, the CDROM changer feature does work with "Quirks:0", but does not work with "Quirks:1".
If there is anything I can do to further narrow it down, please let me know.

Thank you for making such a great device.

**Reply by PetteriAimonen** [reply]
Thanks for the detailed information, I was able to reproduce this problem. I have added it as issue #276.

**Reply by PetteriAimonen** [reply]
Hmm, actually no, it doesn't seem to happen to me or only happens randomly.

@ufkel If it happens reliably for you, could you send me the full zululog.txt when debug is enabled and the volume appears not ready?

**Reply by ufkel** [reply]
@PetteriAimonen Thank you for looking into the issue.
In the attached folder you find:
```
zululog.txt
zuluscsi.ini
```

To comment on the log entries, here is what I did:

1. Start up the Macintosh. ZuluSCSI is the only SCSI device. 
2. CDROM does not appear on the desktop.
3. Start FWB Mounter application.
4. Within FWB Mounter, list all devices. SCSI-ID5 device is listed.
5. Within FWB Mounter, switch from device view to volume view. Volume in SCSI-ID5 is shown as "not ready".
6. Quit the FWB Mounter application.
7. Shut down the Macintosh.


[untitled folder.zip](https://github.com/ZuluSCSI/ZuluSCSI-firmware/files/12281122/untitled.folder.zip)

**Reply by PetteriAimonen** [reply]
Weird, it seems to access a few sectors successfully and `TestUnitReady` also returns good status.

Can you take a log of when the exact same image is named CD5.iso instead? And keep the other .ini settings the same, except comment out IMG0..IMG2.

**Reply by ufkel** [reply]
Attached you find several tests I did. #1 is the one you have asked me for. In each directory you find a description of the test settings and the results, plus the log file  and the ini file.

I stand corrected. The problem seems to be related to the ```System=Mac```setting only. With that setting, neither the ```CDn=anything.iso``` nor the CDROM changer do work. But without ```System=Mac```, both features do work.

Can ```Quirks```be applied for a single device, or do they have to be applied globally? 

[Tests.zip](https://github.com/ZuluSCSI/ZuluSCSI-firmware/files/12287701/Tests.zip)

**Reply by PetteriAimonen** [reply]
Yes, overriding `Quirks=0` for the CD-ROM device could help. Quirks can be specified per-device, while System-setting is global.

It appears that the Mac you have does not like the response it gets for ModeSense of 0x30 Apple vendor page of the CD-ROM drive. I think the value that gets returned currently has originally been taken from a harddrive, so maybe that is confusing it.

My Mac LC III doesn't appear to request ModeSense of the CD-drive at all. In one boot I had it not detect the CD drive, but that was probably some kind of glitch from hot-plugging SD cards because I couldn't reproduce it again.

@ufkel: Which Mac hardware and operating system version are you using?

@aperezbios: Do you happen to have any original Apple SCSI CD drive that could be checked for the proper ModeSense response? It would be easiest done on Linux, I can provide commands if this sounds reasonable.

**Reply by aperezbios** [reply]
@PetteriAimonen i have several. Let me know what you want.

**Reply by ufkel** [reply]
@PetteriAimonen The Mac we are talking about is a SE/30, running System 7.1 pro.

**Reply by PetteriAimonen** [reply]
@aperezbios `sg_inq /dev/sr0`, `sg_modes -avH /dev/sr0` and `sg_modes -vH -p 0x30 /dev/sr0` (where `/dev/sr0` is the CD-ROM device). Currently the vendor ID with apple quirks is set to `MATSHITA CD-ROM CR-8004A`.

**Reply by ufkel** [reply]
Not sure if this is interesting for you guys:
Response for page 0x30 in MAME:
[https://github.com/mamedev/mame/blob/badecba0e3aa372b75c738a4c53b2a7f8332064e/src/devices/bus/nscsi/cd.cpp#L427](url)

**Comment by aperezbios** [comment]
@PetteriAimonen here's a dump from a later (Circa-1998) AppleCD-branded internal 24x CD-ROM drive, model CR-508-C
```
$ sg_modes -vH -p 0x30 /dev/sr0 >> blah
    inquiry cdb: [12 00 00 00 24 00]
    mode sense(10) cdb: [5a 00 30 00 00 00 00 10 00 00]
mode sense(10):
Fixed format, current; Sense key: Illegal Request
Additional sense: Invalid command operation code
>>>>>> try again with a '-6' switch for a 6 byte MODE SENSE command
Illegal request, Invalid opcode, type: sense key + asc,ascq
$ sg_modes -vH -p 0x30 -6 /dev/sr0 >> blah
    inquiry cdb: [12 00 00 00 24 00]
    mode sense(6) cdb: [1a 00 30 00 fc 00]
$ sg_inq /dev/sr0
standard INQUIRY:
  PQual=0  Device_type=5  RMB=1  LU_CONG=0  version=0x02  [SCSI-2]
  [AERC=0]  [TrmTsk=0]  NormACA=0  HiSUP=0  Resp_data_format=2
  SCCS=0  ACC=0  TPGS=0  3PC=0  Protect=0  [BQue=0]
  EncServ=0  MultiP=0  [MChngr=0]  [ACKREQQ=0]  Addr16=0
  [RelAdr=0]  WBus16=0  Sync=1  [Linked=1]  [TranDis=0]  CmdQue=0
    length=36 (0x24)   Peripheral device type: cd/dvd
 Vendor identification: MATSHITA
 Product identification: CD-ROM CR-8024
 Product revision level: 2.0f
```

**Reply by PetteriAimonen** [reply]
Hmm, in `sg_modes -vH -p 0x30 -6 /dev/sr0 >> blah` could I get the blah file also? It seems the output was redirected there.

**Reply by aperezbios** [reply]
Whoops, my bad, @PetteriAimonen. Here's the output:
```
$  sg_modes -vH -p 0x30 -6 /dev/sr0
    inquiry cdb: [12 00 00 00 24 00]
    MATSHITA  CD-ROM CR-8024    2.0f   peripheral_type: cd/dvd [0x5]
    mode sense(6) cdb: [1a 00 30 00 fc 00]
 00     23 00 00 08 00 00 00 00  00 00 08 00 30 16 41 50
 10     50 4c 45 20 43 4f 4d 50  55 54 45 52 2c 20 49 4e
 20     43 20 20 20
```

**Reply by PetteriAimonen** [reply]
Ok, that almost matches the MAME code linked above. Currently ZuluSCSI has one extra 0x00 at the end.
MAME code has length as 0x00, standard page format would have 0x17, and the capture above has 0x16.

@ufkel Can you try this test build? It should match the format captured from actual drive, hopefully it will be accepted by the Mac now. https://github.com/ZuluSCSI/ZuluSCSI-firmware/actions/runs/5819553900

**Reply by ufkel** [reply]
Hi @PetteriAimonen,
I tried the test build, no success. 

- Test 20230810: With ```System=Mac```, the CDROM does not appear on the desktop. ```FWB Mount``` does see the SCSI device, but the volume is "not ready".
- Test 20230810-2: Without ```System=Mac```, the problem now persists (!). This is worse than before. CDROM does not appear on the desktop. ```FWB Mount``` does see the SCSI device, and the volume is "not ready".


[Test 20230810.zip](https://github.com/ZuluSCSI/ZuluSCSI-firmware/files/12311636/Test.20230810.zip)
[Test 20230810-2.zip](https://github.com/ZuluSCSI/ZuluSCSI-firmware/files/12311707/Test.20230810-2.zip)

**Reply by PetteriAimonen** [reply]
That change shouldn't affect behavior without System=Mac, but indeed from the logs something weird is happening. I'll have to study them in depth and try to reproduce the situation here.

**Reply by PetteriAimonen** [reply]
Looks like there was a long-standing bug inherited through SCSI2SD v6 code base, that had been separately fixed in SCSI2SD v5. It didn't affect behavior in most cases.

@ufkel Could you try this build:
https://github.com/ZuluSCSI/ZuluSCSI-firmware/actions/runs/5830174911

Also to be sure, comment out the `Product = "CD-ROM Drive"` line in the config so that the default `CD-ROM CR-8004A` for Mac gets applied. I'm not sure if Mac actually cares, but worth trying.

**Reply by ufkel** [reply]
Hi @PetteriAimonen,
I tried the new test build, here are the results:

- Test 20230811-1: With ```System=Mac```, with ```Product = "CD-ROM Drive"```: CDROM does **not** appear on the desktop. FWB Mounter does see the SCSI device, but the volume is "not ready". The device is identified as "CD-ROM".
- Test 20230811-2: Without ```System=Mac```, with ```Product = "CD-ROM Drive"```: CDROM does appear on the desktop and works OK. 

This behavior is similar to the current stock firmware.

- Test 20230811-3: With ```System=Mac```, without ```Product = "CD-ROM Drive"```: CDROM does **not** appear on the desktop. FWB Mounter does see the SCSI device, but the volume is "not ready". The device is identified as "CD-ROM CR-8004A". 
- Test 20230811-4: Without ```System=Mac```, without ```Product = "CD-ROM Drive"```: CDROM does appear on the desktop and works OK. The device is identified as "ZULUSCSICDROM".

[20230811-1.zip](https://github.com/ZuluSCSI/ZuluSCSI-firmware/files/12323084/20230811-1.zip)
[20230811-2.zip](https://github.com/ZuluSCSI/ZuluSCSI-firmware/files/12323085/20230811-2.zip)
[20230811-3.zip](https://github.com/ZuluSCSI/ZuluSCSI-firmware/files/12323086/20230811-3.zip)
[20230811-4.zip](https://github.com/ZuluSCSI/ZuluSCSI-firmware/files/12323089/20230811-4.zip)

**Reply by PetteriAimonen** [reply, marked as answer]
Weird. I don't currently know what could be causing the problem - the operating system doesn't seem to like the drive with the Mac-specific 0x30 mode sense page, but then accepts the drive that does not have the page.

In any case, it did uncover a bug in the code so at least there was some progress. For now you can use `Quirks=0` on the device to make it work for you.

**Reply by ufkel** [reply]
Hi @PetteriAimonen,

Just to bring the results together: 
What kind of CD-ROM extension/CP are you using? Which OS version do you use on your LC-III?

**Reply by PetteriAimonen** [reply]
For the tests I used OS version 7.5.3. I'm not sure about extensions, there is a file called `System Folder/Extensions/ISO 9660 File Access`. I don't know much about classic Macs, I got the image from @aperezbios.
